File: python/src/seventhplace/merge.py
# SPDX-License-Identifier: Apache-2.0
"""Deep merge utility for layered configuration."""

from __future__ import annotations

from typing import Any


def deep_merge(base: dict[str, Any], override: dict[str, Any]) -> dict[str, Any]:
    """Recursively merge override into base, returning a new dict.

    - Dict values: recurse (preserves sibling keys).
    - Non-dict values: override replaces base.
    - Neither input is mutated.
    """
    result = base.copy()
    for key, value in override.items():
        if isinstance(value, dict) and isinstance(result.get(key), dict):
            result[key] = deep_merge(result[key], value)
        else:
            result[key] = value
    return result

File: typescript/src/merge.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Deep merge utility for layered configuration.
 */

/**
 * Recursively merge override into base, returning a new object.
 *
 * - Object values: recurse (preserves sibling keys).
 * - Non-object values: override replaces base.
 * - Neither input is mutated.
 */
export function deepMerge(
  base: Record<string, unknown>,
  override: Record<string, unknown>,
): Record<string, unknown> {
  const result: Record<string, unknown> = { ...base };

  for (const [key, value] of Object.entries(override)) {
    const baseVal = result[key];
    if (
      typeof value === "object" &&
      value !== null &&
      !Array.isArray(value) &&
      typeof baseVal === "object" &&
      baseVal !== null &&
      !Array.isArray(baseVal)
    ) {
      result[key] = deepMerge(
        baseVal as Record<string, unknown>,
        value as Record<string, unknown>,
      );
    } else {
      result[key] = value;
    }
  }

  return result;
}
